File: android/app/src/main/kotlin/com/sigilauth/app/network/models/Challenge.kt
package com.wagmilabs.sigil.network.models

import com.google.gson.annotations.SerializedName

/**
 * Challenge-related data models per OpenAPI schema.
 *
 * Per D10: uses pictogram_speakable (space-separated).
 *
 * AGPL-3.0 License
 */

/**
 * Challenge request (POST /challenge).
 */
data class ChallengeRequest(
    @SerializedName("fingerprint")
    val fingerprint: String,  // Hex-encoded 64-char fingerprint

    @SerializedName("device_public_key")
    val devicePublicKey: String,  // Base64-encoded compressed P-256 key

    @SerializedName("action")
    val action: Action
)

/**
 * Challenge created response (201 from POST /challenge).
 */
data class ChallengeCreated(
    @SerializedName("challenge_id")
    val challengeId: String,  // UUID

    @SerializedName("fingerprint")
    val fingerprint: String,

    @SerializedName("pictogram")
    val pictogram: List<String>,

    @SerializedName("pictogram_speakable")
    val pictogramSpeakable: String,  // Space-separated per D10

    @SerializedName("expires_at")
    val expiresAt: String  // ISO8601 timestamp
)

/**
 * Challenge response (POST /respond).
 */
data class ChallengeResponse(
    @SerializedName("challenge_id")
    val challengeId: String,

    @SerializedName("device_public_key")
    val devicePublicKey: String,

    @SerializedName("decision")
    val decision: String,  // "approved" | "denied"

    @SerializedName("signature")
    val signature: String,  // Base64-encoded ECDSA signature

    @SerializedName("timestamp")
    val timestamp: String  // ISO8601 timestamp
)

/**
 * Challenge verified response (200 from POST /respond).
 */
data class ChallengeVerified(
    @SerializedName("verified")
    val verified: Boolean,

    @SerializedName("fingerprint")
    val fingerprint: String,

    @SerializedName("pictogram")
    val pictogram: List<String>,

    @SerializedName("pictogram_speakable")
    val pictogramSpeakable: String,  // Space-separated per D10

    @SerializedName("action")
    val action: Action
)

/**
 * Challenge status (GET /v1/auth/challenge/{id}/status).
 */
data class ChallengeStatus(
    @SerializedName("challenge_id")
    val challengeId: String,

    @SerializedName("status")
    val status: String,  // "pending" | "verified" | "denied" | "expired"

    @SerializedName("fingerprint")
    val fingerprint: String? = null,

    @SerializedName("pictogram")
    val pictogram: List<String>? = null,

    @SerializedName("pictogram_speakable")
    val pictogramSpeakable: String? = null,  // Space-separated per D10

    @SerializedName("decision")
    val decision: String? = null,

    @SerializedName("verified_at")
    val verifiedAt: String? = null
)

/**
 * Action context (step-up, login, decrypt, MPA).
 */
data class Action(
    @SerializedName("type")
    val type: String,

    @SerializedName("description")
    val description: String,

    @SerializedName("params")
    val params: Map<String, Any>? = null,

    @SerializedName("details_url")
    val detailsUrl: String? = null  // Optional URL to view action details in external browser
)

File: ios/Sources/Core/Models/Challenge.swift
import Foundation

/// Challenge action context per challenge.json schema
public struct Action: Codable, Equatable, Sendable {
    public let type: String
    public let description: String
    public let params: [String: String]?

    public init(type: String, description: String, params: [String: String]? = nil) {
        self.type = type
        self.description = description
        self.params = params
    }
}

/// Challenge creation request per challenge.json schema
public struct ChallengeRequest: Codable, Equatable, Sendable {
    public let fingerprint: String
    public let devicePublicKey: String
    public let action: Action

    enum CodingKeys: String, CodingKey {
        case fingerprint
        case devicePublicKey = "device_public_key"
        case action
    }

    public init(fingerprint: String, devicePublicKey: String, action: Action) {
        self.fingerprint = fingerprint
        self.devicePublicKey = devicePublicKey
        self.action = action
    }
}

/// Challenge creation response per challenge.json schema
public struct ChallengeCreated: Codable, Equatable, Sendable {
    public let challengeId: String
    public let fingerprint: String
    public let pictogram: [String]
    public let pictogramSpeakable: String
    public let expiresAt: Date

    enum CodingKeys: String, CodingKey {
        case challengeId = "challenge_id"
        case fingerprint
        case pictogram
        case pictogramSpeakable = "pictogram_speakable"
        case expiresAt = "expires_at"
    }
}

/// Challenge push notification payload per challenge.json schema
public struct ChallengeNotification: Codable, Equatable, Sendable {
    public let type: String
    public let challengeId: String
    public let serverId: String
    public let challengeBytes: String
    public let action: Action
    public let serverSignature: String
    public let expiresAt: Date
    public let respondTo: String

    enum CodingKeys: String, CodingKey {
        case type
        case challengeId = "challenge_id"
        case serverId = "server_id"
        case challengeBytes = "challenge_bytes"
        case action
        case serverSignature = "server_signature"
        case expiresAt = "expires_at"
        case respondTo = "respond_to"
    }
}

/// Challenge response submission per challenge.json schema
public struct ChallengeResponse: Codable, Equatable, Sendable {
    public let challengeId: String
    public let devicePublicKey: String
    public let decision: Decision
    public let signature: String
    public let timestamp: Date

    public enum Decision: String, Codable, Sendable {
        case approved
        case rejected
    }

    enum CodingKeys: String, CodingKey {
        case challengeId = "challenge_id"
        case devicePublicKey = "device_public_key"
        case decision
        case signature
        case timestamp
    }

    public init(
        challengeId: String,
        devicePublicKey: String,
        decision: Decision,
        signature: String,
        timestamp: Date
    ) {
        self.challengeId = challengeId
        self.devicePublicKey = devicePublicKey
        self.decision = decision
        self.signature = signature
        self.timestamp = timestamp
    }
}

/// Challenge verification response per challenge.json schema
public struct ChallengeVerified: Codable, Equatable, Sendable {
    public let verified: Bool
    public let fingerprint: String
    public let pictogram: [String]
    public let pictogramSpeakable: String
    public let action: Action?

    enum CodingKeys: String, CodingKey {
        case verified
        case fingerprint
        case pictogram
        case pictogramSpeakable = "pictogram_speakable"
        case action
    }
}
